consul/src/main/java/com/linecorp/armeria/internal/consul/ConsulClient.java

```diff
@@ -0,0 +1,121 @@
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.linecorp.armeria.client.ClientDecoration;
+import com.linecorp.armeria.client.ClientOptions;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.client.WebClientBuilder;
+import com.linecorp.armeria.client.retry.RetryRule;
+import com.linecorp.armeria.client.retry.RetryingClient;
+import com.linecorp.armeria.common.HttpResponse;
+
+/**
+ * A client for accessing to Consul agent API server.
+ */
+public final class ConsulClient {
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private static final ClientOptions retryingClientOptions =
+            ClientOptions.of(ClientOptions.DECORATION.newValue(ClientDecoration.of(
+                    RetryingClient.newDecorator(RetryRule.failsafe(), 3))));
+
+    private final WebClient webClient;
+    private final HealthClient healthClient;
+
+    public ConsulClient(URI uri) {
+        this(uri, null);
+    }
+
+    public ConsulClient(URI uri, @Nullable String token) {
+        final WebClientBuilder builder = WebClient.builder(uri);
+        builder.options(retryingClientOptions);
+        if (token != null) {
+            // TODO(eugene70) test with token
+            builder.addHeader("X-Consul-Token", token);
+        }
+        webClient = builder.build();
+        healthClient = HealthClient.of(this);
+    }
+
+    /**
+     * Returns an {@link ObjectMapper} that is used to encode and decode Consul requests and responses.
+     */
+    public ObjectMapper getObjectMapper() {
+        return objectMapper;
+    }
+
+    /**
+     * Registers a service to Consul Agent with service ID.
+     *
+     * @param serviceId a service ID that identifying a service
+     * @param serviceName a service name to register
+     * @param endpoint an endpoint of service to register
+     * @param check a check for the service
+     * @return a {@link CompletableFuture} that will be completed with the registered service ID
+     */
+    public HttpResponse register(String serviceId, String serviceName, Endpoint endpoint,
+                                 @Nullable Check check) {
+        return AgentServiceClient.of(this)
+                                 .register(serviceId, serviceName, endpoint.host(), endpoint.port(), check);
+    }
+
+    /**
+     * De-registers a service to Consul Agent.
+     *
+     * @param serviceId a service ID that identifying a service
+     */
+    public HttpResponse deregister(String serviceId) {
+        return AgentServiceClient.of(this).deregister(serviceId);
+    }
+
+    /**
+     * Get registered endpoints with service name from Consul agent.
+     */
+    public CompletableFuture<List<Endpoint>> endpoints(String serviceName) {
+        return CatalogClient.of(this).endpoints(serviceName);
```

Comment:
Can we make `CatalogClient ` once in the constructor and reuse it?

Reply:
Yep.
`AgentServiceClient` also.